Add GA workflow to update AtoM's Vagrant box

jraddaoui opened the pull request:
Add a manually triggered Github actions workflow, accepting `version`
and `description` as inputs, to build and upload AtoM's Vagrant box. Use
a Ruby script to upload the box with Hashicorp's Ruby client for the
Vagrant Cloud API.

Refs https://trello.com/c/vMINa4M8/323-atom-vagrant-box-automated-build-and-upload.

jraddaoui commented on .github/workflows/vagrant-box-atom.yml:
```diff
@@ -0,0 +1,36 @@
+      version:
+        description: Version
+        required: true
+      description:
+        description: Description
+        required: true
```
I'd love to see a textarea input in here but it's currently not possible. The easiest solution I found is to add `<br/>` directly  in the input and those are represented in Vagrant Cloud's GUI as expected, but they break the Markdown support. We could try to add some extra formatting in the Ruby script to accept other linebreaks (like `\n`) but I'm not sure if it worths the effort.

sevein replied:
One potential improvement for the future is to maintain a CHANGELOG.md document (there are many changelog parsers or plain markdown parsers) in this repo that is read by this workflow to extract the description field from there - it's somewhat more reproducible too since the build inputs will be versioned.

jraddaoui commented on .github/workflows/vagrant-box-atom.yml:
```diff
@@ -0,0 +1,36 @@
+jobs: 
+  vagrant-box-atom:
+    name: Build and upload
+    runs-on: macos-10.15
```
MacOS required for nested virtualization:

https://github.com/actions/virtual-environments/issues/183
https://github.com/actions/virtual-environments/issues/433

It also comes with all the required dependencies (Virtualbox, Vagrant, Packer, Bundler).

jraddaoui commented on .github/workflows/vagrant-box-atom.yml:
```diff
@@ -0,0 +1,36 @@
+        bundle install
+        ruby upload.rb \
+          '${{ github.workspace }}/atom-vagrant-${{ github.event.inputs.version }}.box' \
+          '${{ secrets.VAGRANT_CLOUD }}' \
```
Already generated and added to the repo.

jraddaoui commented on tools/vagrant-box-atom/upload.rb:
```diff
@@ -0,0 +1,59 @@
+version = ARGV[2]
+description = ARGV[3]
+
+client = VagrantCloud::Client.new(access_token: token)
+
+client.box_version_create(
+    username: ORG,
+    name: BOX,
+    version: version,
+    description: description
+)
+
+client.box_version_provider_create(
+    username: ORG,
+    name: BOX,
+    version: version,
+    provider: PROVIDER
+)
+
+upload_url = client.box_version_provider_upload(
+    username: ORG,
+    name: BOX,
+    version: version,
+    provider: PROVIDER
+)
```
If anything fails here (authentication, version already exists, etc.) the script ends with error. However, this may leave residual data in Vagrant Cloud that won't be considered in a a new run. For example, if the box fails to upload below and the workflow is triggered again with the same version it will fail because the version already exists.

We could consider those cases in the script but I'm not sure if we should, as it may lead to versions overwrite. We could document the situation so, in case of failure, if the version has been created it has to be deleted from Vagrant Cloud's GUI before trying again with the same version.

jraddaoui replied:
I guess we could check if the version is fully released to avoid that overwrite, but I still have my doubts about if it's worth it, feedback is welcomed.

sevein replied:
For error handling, maybe one option is a workflow job with a [status condition](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#job-status-check-functions) that does the necessary clean-up?

jraddaoui replied:
Thanks @sevein, I think it may be easier directly in the script, but it is an enhancement that could be added later on. For now I'll add a note in the internal wiki about failure/retry.

sevein commented on tools/vagrant-box-atom/upload.rb:
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/ruby
```
If it's not a big deal I'd probably change it to:

    #!/usr/bin/env ruby

That ensures compatibility with other platforms where the binary sits in other locations.

jraddaoui commented on tools/vagrant-box-atom/upload.rb:
```diff
@@ -0,0 +1,59 @@
+require 'vagrant_cloud'
+
+ORG = 'artefactual'
+BOX = 'atom'
```
At the moment it's all named for AtoM's Vagrant box, but we could accept this as a script argument and reuse the same script for Archivematica's Vagrant box, if we want to implement a similar workflow.

jraddaoui commented on tools/vagrant-box-atom/upload.rb:
```diff
@@ -0,0 +1,59 @@
+BOX = 'atom'
+PROVIDER = 'virtualbox'
+
+path = ARGV[0]
+token = ARGV[1]
+version = ARGV[2]
+description = ARGV[3]
```
We could use [Ruby's option parser](https://ruby-doc.org/stdlib-2.1.3/libdoc/optparse/rdoc/OptionParser.html) to improve the script arguments/options management and document the script usage if needed.

sevein commented on tools/vagrant-box-atom/Gemfile.lock:
```diff
@@ -0,0 +1,18 @@
+
+PLATFORMS
+  ruby
+  x86_64-linux
```
I guess it doesn't really matter since you're not publishing anything but shouldn't this platform be a "darwin" something? Feel free to ignore.

jraddaoui replied:
TBH, I have no idea, this is what `bundle install` created and seems to be working in both Linux and MacOS.

sevein reviewed (approved):
This is useful! A future improvement could be to push a git tag as soon as you're done with the workflow and it executed successfully, if we wanted to track that sort of thing in this repository.
A common git tag naming pattern in monorepos is `subproject/version`, e.g. `vagrant-box-atom/${version}`, e.g. [`gopls/v0.6.10`](https://github.com/golang/tools/releases/tag/gopls%2Fv0.6.10). I don't think we really need another action just to push a git tag, but there's [action-push-tag](https://github.com/actions-ecosystem/action-push-tag) for reference.

I'm aware we don't have a lot of documentation in this repo, but a tools/vagrant-box-atom/README.md shortly describing its purpose would be perfect.

jraddaoui commented:
Thanks @sevein!

About the changelog and the tag, I'm not sure due to how this repo is managed/named and all the other repositories that are involved, a new Vagrant box may be needed for many reasons and most of them won't involve this repository.

jraddaoui committed:
Add GA workflow to update AtoM's Vagrant box

Add a manually triggered Github actions workflow, accepting `version`
and `description` as inputs, to build and upload AtoM's Vagrant box. Use
a Ruby script to upload the box with Hashicorp's Ruby client for the
Vagrant Cloud API.
